### ISSUE TITLE
^2.4.0 when reinstalled pulls 2.7.0 which has breaking api changes

### ISSUE THREAD

**Issue by phillipDarby** [gregnb/mui-datatables#813]
## Expected Behavior
the api should remain the same on a minor version increase otherwise change to a major version if the api is different

## Current Behavior
with select rows disabled it enables when updated to 2.7.0 and the way i have it coded works perfect on 2.4.0, there are other bugs too like now on some pages i get filter undefined but works fine in 2.4.0 

## Steps to Reproduce (for bugs)

1. have 2.4.0 mui-datatables installed
2. reinstall npm packages it updates to 2.7.0 which is a minor update
3. the api has breaking changes it should be a major update if the api has breaking changes
4. I do not know at what version it broke but api changes should of bumped it to 3.0.0 i have to keep a backup copy of 2.4.0 because using ^2.4.0 will update the breaking api changes.

## Your Environment

| Tech         | Version |
|--------------|---------|
| Material-UI  |    ^3.9.2     |
| MUI-datatables  |     ^2.4.0    |
| React        |    ^16.8.6     |
| browser      |     google chrome    |
| etc          |         |

**Comment by gabrielliwerant** [comment]
Hello @phillipDarby. In order to address your issues, I would need specific tickets with steps to reproduce and/or code examples I can run. If you can provide that, I will look into the issues.

New features are allowable in minor version updates, as long as they are backwards compatible, which I do my best to ensure. It's always possible that I made a mistake, which is why I would like to learn more about your issue.

I do think it's worth pointing out though, that in principle, there is no such thing as any version update, even patches, that can guarantee that your code will still work before and after. It is always possible, for example, for code to be written that unknowingly takes advantage of a bug, and will thus break even during a patch.

So, one specific example I know of is that many users have been using `false` as a value to send in to the `selectableRows` option. I don't know the full history of this feature, but I think that it was unintentional to allow use of `false` for an option which is listed in the documentation as an enum. So, at some point, the incorrect option was working because the code either allowed it, or it was allowed incidentally, by accident. This kind of issue should be fixed, as it is not the correct API, and can lead to other bugs and unintended consequences. However, on fixing it, anyone improperly using `false` as an option will see their code break. But it's important to follow the docs as best as possible.

**Comment by phillipDarby** [comment]
here is an example of one of the components 
`import React, { Component } from "react";
import withAuth from "../../auth/withAuth";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import IconButton from "@material-ui/core/IconButton";
import Button from "@material-ui/core/Button";
import MuiThemeProvider from "@material-ui/core/styles/MuiThemeProvider";
import Visibility from "@material-ui/icons/Visibility";
import Avatar from "@material-ui/core/Avatar";
import Chip from "@material-ui/core/Chip";
import Tooltip from "@material-ui/core/Tooltip";
import { withStyles } from "@material-ui/core/styles";
import Green from "@material-ui/core/colors/green";
import Red from "@material-ui/core/colors/red";
import MUIDataTable from "mui-datatables";
import Warning from "@material-ui/icons/Warning";
import {
 
  getMuiTheme
} from "./HelperFunctions/helpers";

const ContractedChip = withStyles({
  root: {
    backgroundColor: Green[600],
    color: "white",
    fontWeight: "700",
    margin: 3
  }
})(Avatar);

const NonContractedChip = withStyles({
  root: {
    backgroundColor: Red[600],
    color: "white",
    fontWeight: "700",
    margin: 3
  }
})(Avatar);

/**
 * NOTE: always check is mounted before setting the state
 */
class MyClientsContainer extends Component {
  state = {
    data: [],
    page: 0,
    count: 0,
    rowsPerPage: 10,
    filterList: [[], [], [], [], [], [], [], []],
    columns: []
  };

  _baseStore = "myClientsTab-";
  _isMounted = false;

  checkFilters = () => {
      let bool = false;
    this.state.columns.forEach(function(column){
      if(column.display === "false"){bool = true;}
    });
    if (bool) {

      return (
        <Tooltip title="Not All Columns are active">
          <IconButton style={{ color: Red[600] }}>
            <Warning />
          </IconButton>
        </Tooltip>
      );
    }
  };

  componentDidMount() {
    this._isMounted = true;

    this.setState({
      rowsPerPage: localStorage.getItem(this._baseStore + "NumberRows") !== null
        ? parseInt(localStorage.getItem(this._baseStore + "NumberRows"))
        : 10,

        filterList: localStorage.getItem(this._baseStore + "Filter") !== null
        ? JSON.parse(localStorage.getItem(this._baseStore + "Filter"))
        : [[], [], [], [], [], [], [], [], []],

        columns: localStorage.getItem(this._baseStore + "Columns") !== null
        ? JSON.parse(localStorage.getItem(this._baseStore + "Columns"))
        :[]
    });

    this.getData();
    
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  /**
   * TODO: Enable this once api is done for this call...
   */
  getData = async () => {
    let token = localStorage.getItem("token");
    await fetch(`api/clientLocation/getMyClientsTabData`, {
      headers: {
        Authorization: "Bearer " + token,
        "Content-Type": "application/json"
      },
      method: "post"
    })
      .then(response => {
        if (!response.ok) {
          this.props.unauthorized();
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(data => {
        if (this._isMounted) {
          /** HACK: this is only way I can find to download CSV properly need to change to field that may have commas in it */
          var dat = data.map(val => {
            val[2] = val[2].replace(",", " ");
            return val;
          });
          this.setState({
            data: dat,
            count: data.length
          });
        }
      })
      .then(() => {
        var test = this.state.data.filter(rec => {
          return rec[1] === this.state.filterList[1];
        });

        if (this._isMounted && test !== null) {
          this.setState({ count: test.length });
        }
      })
      .catch(error => {
        console.log(error);
      });
  };

  handleChangeRowsPerPage = event => {
    localStorage.setItem(this._baseStore + "NumberRows", event.target.value);
    this.setState({ rowsPerPage: event.target.value });
  };

  handleChangePage = (event, page) => {
    this.setState({ page: page });
  };

  render() {
    const { data, page, count, rowsPerPage } = this.state;
    const options = {
      print: false,
      filterType: "multiselect",
      selectableRows: false,
      filter: true,
      responsive: "scroll",
      serverSide: false,
      download: this.props.users.permissions.auditing,
      downloadOptions: { filename: "MyClients.csv", separator: "," },
      textLabels: {
        pagination: {
          next: "Next Page",
          previous: "Previous Page",
          rowsPerPage: "Clients per page:",
          displayRows: "of"
        },
        body: {
          noMatch: "Sorry, you have no clients assigned to you",
          toolTip: "Sort"
        }
      },
      count: count,
      page: page,
      rowsPerPage: rowsPerPage,

      onFilterChange: (changedColumn, list) => {
        localStorage.setItem(this._baseStore + "Filter", JSON.stringify(list));
        this.setState({ filterList: list });
      },
      onColumnViewChange: (changedColumn, action) => {
        let display = action === "remove" ? "false" : "true";
        let tColumns = this.state.columns;
        tColumns.map((column) => {
          if(changedColumn === column.name){
            column.display = display;
          }
        });
        this.setState({columns: tColumns});
        localStorage.setItem(this._baseStore + "Columns", JSON.stringify(tColumns));

      },

      onColumnSortChange: (changedColumn, direction) => {
        let dir = direction === "ascending" ? "asc" : "desc";
        let tColumns = this.state.columns;
        tColumns.map((column) => {
          column.sortDirection = changedColumn === column.name ? dir : null;
        });
        this.setState({columns: tColumns});
        localStorage.setItem(this._baseStore + "Columns", JSON.stringify(tColumns));

      },
      onTableInit: (action, tableState) => {
        let empty = localStorage.getItem(this._baseStore + "Columns") === null;
        
        if(empty){
          localStorage.setItem(this._baseStore + "Columns",JSON.stringify(tableState.columns));
        }
      },

      onTableChange: (action, tableState) => {
        if(action === "propsUpdate") {
          for(let i = 0; i < this.state.columns.length; i++){
            tableState.columns[i].sortDirection = this.state.columns[i].sortDirection;
            tableState.columns[i].display = this.state.columns[i].display;
          }
          tableState.filterList = this.state.filterList;
          tableState.rowsPerPage = this.state.rowsPerPage;
          tableState.count = tableState.displayData.length
        }

        if(action === "filterChange")
        {
          tableState.count = tableState.displayData.length
        }
      },

      onChangePage: currentPage => {
        this.setState({ page: currentPage });
      },

      onChangeRowsPerPage: numberOfRows => {
        localStorage.setItem(this._baseStore + "NumberRows", numberOfRows);
        this.setState({ rowsPerPage: numberOfRows });
      },

      customToolbar: tableMeta => {
        return <React.Fragment>{this.checkFilters()}</React.Fragment>;
      }
    };
    const columns = [
      {
        name: "Action",
        options: {
          filter: false,
          download: false,
          customBodyRender: (value, tableMeta, updateValue) => {
            return (
              //<Tooltip title="View">

              <Button
                size="small"
                onClick={() =>
                  this.props.history.push(`/clients/location/${value}`)
                }
              >
                <Visibility color="primary" aria-label="View" />
              </Button>

              //</Tooltip>
            );
          }
        }
      },
      {
        name: "Contracted",
        options: {
          customFilterListRender: v => `Contracted: ${v}`,
          sort: true,
          customBodyRender: (value, tableMeta, updateValue) => {
            if (value === "Contracted") {
              return (
                <Chip
                  variant="outlined"
                  avatar={<ContractedChip sizes="sm">C</ContractedChip>}
                />
              );
            }
            if (value === "Non-Contracted") {
              return (
                <Chip
                  variant="outlined"
                  avatar={<NonContractedChip sizes="sm">NC</NonContractedChip>}
                />
              );
            }
          }
        }
      },
      {
        name: "Name",
        options: {
          filter: false
        }
      },
      {
        name: "Street Address",
        options: {
          filter: false
        }
      },
      {
        name: "City",
        options: {
          filter: false
        }
      },
      {
        name: "State",
        options: {
          filter: false
        }
      },
      {
        name: "Zip",
        options: {
          filter: false
        }
      },
      {
        name: "Phone",
        options: {
          filter: false
        }
      }
    ];
    return (
      <MuiThemeProvider theme={getMuiTheme}>
        <MUIDataTable
          title={"My Clients"}
          data={data}
          columns={columns}
          options={options}
        />
      </MuiThemeProvider>
    );
  }
}

const mapStateToProps = state => {
  return {
    authenticated: state.users.authenticated,
    ui: state.ui,
    users: state.users
  };
};

export default withRouter(
  withAuth(connect(mapStateToProps)(MyClientsContainer))
);`

**Comment by phillipDarby** [comment]
here is an example code:
```
const options = {
      print: false,
      filterType: "multiselect",
      selectableRows: false,
      filter: true,
      responsive: "scroll",
      serverSide: false,
      download: this.props.users.permissions.auditing,
      downloadOptions: { filename: "MyClients.csv", separator: "," },
      textLabels: {
        pagination: {
          next: "Next Page",
          previous: "Previous Page",
          rowsPerPage: "Clients per page:",
          displayRows: "of"
        },
        body: {
          noMatch: "Sorry, you have no clients assigned to you",
          toolTip: "Sort"
        }
      },
      count: count,
      page: page,
      rowsPerPage: rowsPerPage,

      onFilterChange: (changedColumn, list) => {
        localStorage.setItem(this._baseStore + "Filter", JSON.stringify(list));
        this.setState({ filterList: list });
      },
      onColumnViewChange: (changedColumn, action) => {
        let display = action === "remove" ? "false" : "true";
        let tColumns = this.state.columns;
        tColumns.map((column) => {
          if(changedColumn === column.name){
            column.display = display;
          }
        });
        this.setState({columns: tColumns});
        localStorage.setItem(this._baseStore + "Columns", JSON.stringify(tColumns));

      },

      onColumnSortChange: (changedColumn, direction) => {
        let dir = direction === "ascending" ? "asc" : "desc";
        let tColumns = this.state.columns;
        tColumns.map((column) => {
          column.sortDirection = changedColumn === column.name ? dir : null;
        });
        this.setState({columns: tColumns});
        localStorage.setItem(this._baseStore + "Columns", JSON.stringify(tColumns));

      },
      onTableInit: (action, tableState) => {
        let empty = localStorage.getItem(this._baseStore + "Columns") === null;
        
        if(empty){
          localStorage.setItem(this._baseStore + "Columns",JSON.stringify(tableState.columns));
        }
      },

      onTableChange: (action, tableState) => {
        if(action === "propsUpdate") {
          for(let i = 0; i < this.state.columns.length; i++){
            tableState.columns[i].sortDirection = this.state.columns[i].sortDirection;
            tableState.columns[i].display = this.state.columns[i].display;
          }
          tableState.filterList = this.state.filterList;
          tableState.rowsPerPage = this.state.rowsPerPage;
          tableState.count = tableState.displayData.length
        }

        if(action === "filterChange")
        {
          tableState.count = tableState.displayData.length
        }
      },

      onChangePage: currentPage => {
        this.setState({ page: currentPage });
      },

      onChangeRowsPerPage: numberOfRows => {
        localStorage.setItem(this._baseStore + "NumberRows", numberOfRows);
        this.setState({ rowsPerPage: numberOfRows });
      },

      customToolbar: tableMeta => {
        return <React.Fragment>{this.checkFilters()}</React.Fragment>;
      }
    };
```

**Comment by phillipDarby** [comment]
so selectable rows should be `'none'` instead of `false`

**Comment by gabrielliwerant** [comment]
Yes, that is correct regarding `selectableRows` (`'none'` is correct). Does it fix your issue if you make that adjustment?

Perhaps I should have been more clear in the release notes that I was plugging that particular hole. Or, since I knew people were using it, perhaps it would have been even better to add console warnings/errors for the incorrect option. I will consider adding that to help people if this issue comes up again.

**Comment by phillipDarby** [comment]
yes that fixes that issue the other one about filter being undefined i suspect has to do with this area
```
 componentDidMount() {
    this._isMounted = true;

    this.setState({
      rowsPerPage: localStorage.getItem(this._baseStore + "NumberRows") !== null
        ? parseInt(localStorage.getItem(this._baseStore + "NumberRows"))
        : 10,

        filterList: localStorage.getItem(this._baseStore + "Filter") !== null
        ? JSON.parse(localStorage.getItem(this._baseStore + "Filter"))
        : [[], [], [], [], [], [], [], [], []],

        columns: localStorage.getItem(this._baseStore + "Columns") !== null
        ? JSON.parse(localStorage.getItem(this._baseStore + "Columns"))
        :[]
    });

    this.getData();
    
  }
```

or this area

```
{
        name: "Contracted",
        options: {
          customFilterListRender: v => `Contracted: ${v}`,
          sort: true,
          customBodyRender: (value, tableMeta, updateValue) => {
            if (value === "Contracted") {
              return (
                <Chip
                  variant="outlined"
                  avatar={<ContractedChip sizes="sm">C</ContractedChip>}
                />
              );
            }
            if (value === "Non-Contracted") {
              return (
                <Chip
                  variant="outlined"
                  avatar={<NonContractedChip sizes="sm">NC</NonContractedChip>}
                />
              );
            }
          }
        }
      },
```

**Comment by phillipDarby** [comment]
or this area on my fetch 
```
.then(response => {
        if (!response.ok) {
          this.props.unauthorized();
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(data => {
        if (this._isMounted) {
          /** HACK: this is only way I can find to download CSV properly need to change to field that may have commas in it */
          var dat = data.map(val => {
            val[2] = val[2].replace(",", " ");
            return val;
          });
          this.setState({
            data: dat,
            count: data.length
          });
        }
      })
      .then(() => {
        var test = this.state.data.filter(rec => {
          return rec[1] === this.state.filterList[1];
        });

        if (this._isMounted && test !== null) {
          this.setState({ count: test.length });
        }
      })
      .catch(error => {
        console.log(error);
      });
```
